### PR TITLE
8299338: AssertionError in ResponseSubscribers$HttpResponseInputStream::onSubscribe

### PR DESCRIPTION
**Pull request by dfuch**
This fix revisits an assertion that has been observed failing in ResponseSubscribers.HttpResponseInputStream.

The HttpResponseInputStream has the logic to wait until a buffer has been taken out of the queue before requesting a new one. Therefore there should at most be one byte buffer in the queue, except in the case of error (or asychronous close), where a LAST_LIST sentinel is inserted in the queue to unblock the consumer of the input stream, which might be blocked in queue.take().

The  HttpResponseInputStream thus asserts, when processing a subscription, that the remaining capacity of the queue should be greater than 1 (unless already closed), to ensure that there will be room for the LAST_LIST sentinel. However, in case of asynchronous shutdown of the executor, it's possible that the subscriber will be marked failed and the LAST_LIST sentinel inserted into the queue before/at the same time that the subscription is processed.

This fix proposes to relax the assertion to only fire if closed == false and failed == null and capacity <= 1 when processing the subscription

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299338](https://bugs.openjdk.org/browse/JDK-8299338): AssertionError in ResponseSubscribers$HttpResponseInputStream::onSubscribe

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12677/head:pull/12677` \
`$ git checkout pull/12677`

Update a local copy of the PR: \
`$ git checkout pull/12677` \
`$ git pull https://git.openjdk.org/jdk pull/12677/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12677`

View PR using the GUI difftool: \
`$ git pr show -t 12677`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12677.diff">https://git.openjdk.org/jdk/pull/12677.diff</a>

</details>

### CODE REVIEW

**Comment by dfuch** [comment]
@jaikiran made me realise that there is a code path where `register()` might end up in a call to `HttpSubscriberWrapper.onError()`, after the subscriber was markSubscribed() but before `onSubscribe()` was called on the wrapped subscriber. This results in having `onError()` called on the wrapped subscriber before  `onSubscribe()`, which is the very thing that the `HttpSubscriberWrapper` class is trying to prevent. 

I have slightly modified `HttpSubscriberWrapper::onSubscribe` to cater for this, by calling `tryRegister` before calling `markSusbcribed`, but still within the subscriptionLock. Now if register() ends up calling `onError`, the subscriber will not be marked subscribed yet, causing `propagateError` to behave correctly. The real subscription will then be cancelled, which is what we want.

Thanks for the good analysis @jaikiran!

File: src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java [hunk 1]
```diff
@@ -567,20 +574,28 @@ public void registerSubscriber(HttpBodySubscriberWrapper<?> subscriber) {
...
                 }
             }
         }
         subscriber.onError(selmgr.selectorClosedException());
+        return false;
     }
 
-    public void unregisterSubscriber(HttpBodySubscriberWrapper<?> subscriber) {
+    /**
+     * Remove the given subscriber to the subscribers list.
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java:586]: Typo - "to" should have been "from"

**Review by jaikiran — APPROVED** [review]
Thank you for these changes, Daniel. This looks good to me.

I see that now if the tryRegister() doesn't actually register the subscriber and instead the onError() of the subscriber gets called, then in the propagateError() we first mark the subscriber subscribed and then call the onError on the underlying subscriber. So once tryRegister() returns, on the next line the call to markSubscribed() will return false and we then cancel the subscription. So this looks fine to me.

Happy to see this intermittent test failure fixed.

File: src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java (hunk 1 above)
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java:586]:
```suggestion
     * Remove the given subscriber from the subscribers list.
```

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**